torchtitan/train.py

```diff
@@ -518,6 +518,10 @@ def train_step(
             "n_tokens_seen": global_ntokens_seen,
             "lr": lr,
         }
+        for mp in self.model_parts:
+            if mp_metrics := mp.get_extra_metrics(self.parallel_dims):
+                extra_metrics.update(mp_metrics)
```

Comment:
This approach seems quite fragile. What happens if two modules return metrics with the same keys? This could lead to silent overwriting of data.

Reply:
Yes, I have the same concern, but was not sure of how to handle this case, apart from `warning`.  Was hoping to have a discussion about that here

Reply:
This is why I was considering adding the other objects to the `get_extra_metrics` signature above, so you could adjust reporting based on PP scheduler, etc, also.

The logic is going to get pretty complex if a user's `get_extra_metrics` needs to handle all possible sharding cases, but that also seems unavoidable, and at least some API like this could make the path to custom metrics less painful.